From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 12 Dec 2024 02:58:34 -0500
Subject: [PATCH] Cache ShapePairKey hash

License: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
Gale - https://galemc.org

The JMH benchmark of this patch can be found in SunBox's `RecordHashCode`

diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index f6dcd981b6bddadb7618f022908cbbad67095e1b..8a7b1c5444236865a327c59e59d7bc3b69835f0c 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -690,7 +690,20 @@ public class Block extends BlockBehaviour implements ItemLike {
     }
     // CraftBukkit end
 
-    record ShapePairKey(VoxelShape first, VoxelShape second) {
+    // Gale start - cache ShapePairKey hash
+    static class ShapePairKey {
+
+        private final VoxelShape first;
+        private final VoxelShape second;
+        private final int hash;
+
+        private ShapePairKey(VoxelShape first, VoxelShape second) {
+            this.first = first;
+            this.second = second;
+            this.hash = System.identityHashCode(this.first) * 31 + System.identityHashCode(this.second);
+        }
+        // Gale end - cache ShapePairKey hash
+
         @Override
         public boolean equals(Object other) {
             return other instanceof Block.ShapePairKey shapePairKey && this.first == shapePairKey.first && this.second == shapePairKey.second;
@@ -698,7 +711,7 @@ public class Block extends BlockBehaviour implements ItemLike {
 
         @Override
         public int hashCode() {
-            return System.identityHashCode(this.first) * 31 + System.identityHashCode(this.second);
+            return this.hash; // Gale - cache ShapePairKey hash
         }
     }
 
